contracts/protocol/lendingpool/LendingPoolCollateralManagerWithReserve.sol

```diff
@@ -0,0 +1,349 @@
+        address(vars.collateralAtoken)
+      );
+      if (
+        currentAvailableCollateral < vars.maxCollateralToLiquidate.sub(vars.collateralReserveAmount)
+      ) {
+        return (
+          uint256(Errors.CollateralManagerErrors.NOT_ENOUGH_LIQUIDITY),
+          Errors.LPCM_NOT_ENOUGH_LIQUIDITY_TO_LIQUIDATE
+        );
+      }
+    }
+
+    debtReserve.updateState();
+
+    if (vars.userVariableDebt >= vars.actualDebtToLiquidate) {
+      IVariableDebtToken(debtReserve.variableDebtTokenAddress).burn(
+        user,
+        vars.actualDebtToLiquidate,
+        debtReserve.variableBorrowIndex
+      );
+    } else {
+      // If the user doesn't have variable debt, no need to try to burn variable debt tokens
+      if (vars.userVariableDebt > 0) {
+        IVariableDebtToken(debtReserve.variableDebtTokenAddress).burn(
+          user,
+          vars.userVariableDebt,
+          debtReserve.variableBorrowIndex
+        );
+      }
+      IStableDebtToken(debtReserve.stableDebtTokenAddress).burn(
+        user,
+        vars.actualDebtToLiquidate.sub(vars.userVariableDebt)
+      );
+    }
+
+    debtReserve.updateInterestRates(
+      debtAsset,
+      debtReserve.aTokenAddress,
+      vars.actualDebtToLiquidate,
+      0
+    );
+
+    if (receiveAToken) {
+      vars.liquidatorPreviousATokenBalance = IERC20(vars.collateralAtoken).balanceOf(msg.sender);
+      vars.collateralAtoken.transferOnLiquidation(
+        user,
+        msg.sender,
+        vars.maxCollateralToLiquidate.sub(vars.collateralReserveAmount)
+      );
+
+      if (vars.liquidatorPreviousATokenBalance == 0) {
+        DataTypes.UserConfigurationMap storage liquidatorConfig = _usersConfig[msg.sender];
+        liquidatorConfig.setUsingAsCollateral(collateralReserve.id, true);
+        emit ReserveUsedAsCollateralEnabled(collateralAsset, msg.sender);
+      }
+    } else {
+      collateralReserve.updateState();
+      collateralReserve.updateInterestRates(
+        collateralAsset,
+        address(vars.collateralAtoken),
+        0,
+        vars.maxCollateralToLiquidate.sub(vars.collateralReserveAmount)
+      );
+
+      // Burn the equivalent amount of aToken, sending the underlying to the liquidator
+      vars.collateralAtoken.burn(
+        user,
+        msg.sender,
+        vars.maxCollateralToLiquidate.sub(vars.collateralReserveAmount),
+        collateralReserve.liquidityIndex
+      );
+    }
+
+    if (vars.collateralReserveAmount > 0) {
+      vars.collateralAtoken.transferOnLiquidation(
+        user,
+        vars.collateralAtoken.RESERVE_TREASURY_ADDRESS(),
+        vars.collateralReserveAmount
+      );
+    }
+
+    // If the collateral being liquidated is equal to the user balance,
+    // we set the currency as not being used as collateral anymore
+    if (vars.maxCollateralToLiquidate == vars.userCollateralBalance) {
+      userConfig.setUsingAsCollateral(collateralReserve.id, false);
```

Comment:
be careful with this one here, this is the code path that got Agave drained on a re-entrant token.
Let's think a bit more about how to make this safer.

The agave got hacked was:
- collateral has re-entrancy token
- hacker deposit collateral, borrow, and then call liquidate on himself
- liquidate transfer re-entrant token to hacker
- hacker flash deposit more collateral
- hacker borrow another token
- liquidate callback complete, mark isUsingCollateral as false
- hacker withdraw collateral

rinse and repeat against every agave market

Reply:
Aave is specifically does not support re-entrant tokens.
Fixing it in this one place will not guarantee that re-entrant tokens will become safe to use.
We need to make sure that the added tokens are not re-entrant.

Reply:
Yes, taking the Agave fix is a good idea, thank you!